### PR TITLE
[ZEPPELIN-3014] NPE bug fix and Error message enhancement with Kylin Interpreter

### PR DESCRIPTION
**Pull request by jinxliu**
### What is this PR for?
A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - https://zeppelin.apache.org/contribution/contributions.html

### What type of PR is it?
Bug Fix

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3014

### How should this be tested?
*  Setup Travis CI as described on https://zeppelin.apache.org/contribution/contributions.html#continuous-integration
* Use existing unit tests in kylin module.

#### before: NPE when result set is empty
image
#### after: no NPE when result set is empty, just an empty table
image

#### before: when query fails, only error code is returned, no error message
image
#### after: when query fails, both error code and error message are displayed to users
image

### Questions:
* Does the licenses files need update? No.
* Is there breaking changes for older versions? No.
* Does this needs documentation? No.

### CODE REVIEW

**Comment by jinxliu** [comment]
@zjffdu @felixcheung can you please review? I have rebased and created this new pull request. [was https://github.com/apache/zeppelin/pull/2634]. Thanks for your time!

**Comment by zjffdu** [comment]
Thanks @jinxliu  does kylin has some utility class for integration test ? Very few people in community has experience of kylin, it is better to have test for the bug fix and avoid regression

**Comment by jinxliu** [comment]
@zjffdu sure, kylin interpreter has unit test class, and I will add the test when result set is empty.

**Comment by jinxliu** [comment]
@zjffdu please review. Thanks.

File: kylin/src/main/java/org/apache/zeppelin/kylin/KylinErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  public String toJson() {
+    return gson.toJson(this);
+  }
+
+  public static KylinErrorResponse fromJson(String json) {
+    try {
+      return gson.fromJson(json, KylinErrorResponse.class);
+    } catch (JsonSyntaxException ex) {
+      return null;
```
Comment (zjffdu) [inline kylin/src/main/java/org/apache/zeppelin/kylin/KylinErrorResponse.java]: Is this acceptable for kylin to ignore the exception and return null ?

Comment (jinxliu) [reply kylin/src/main/java/org/apache/zeppelin/kylin/KylinErrorResponse.java]: yes, it is acceptable. It is known that when 401 error, the response is in xml format and not json, in this case, just return it as null json, and handle it in KylinInterpreter::executeQuery.

**Comment by zjffdu** [comment]
Thanks @jinxliu wait for CI pass and will merge if no more comments

**Comment by jinxliu** [comment]
@zjffdu the CI checks seems to get stuck, even after rebase.

**Comment by jinxliu** [comment]
@zjffdu Thanks for your comments and review. There seems something wrong with validating pull request.

File: kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java [hunk 2]
```diff
@@ -166,28 +165,42 @@ public String getSQL(String cmd) {
...
+        }
 
-    String output;
-    logger.info("Output from Server .... \n");
-    while ((output = br.readLine()) != null) {
-      logger.info(output);
-      sb.append(output).append('\n');
+        return new InterpreterResult(InterpreterResult.Code.ERROR, errorMessage.toString());
+      }
+    } catch (NullPointerException | IOException e) {
```
Comment (felixcheung) [inline kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java]: NullPointerException shouldn't happen then, if kylinErrorResponse == null is checked?

Comment (jinxliu) [reply kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java]: @felixcheung yes, NPE shouldn't happen now.

File: kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java [hunk 3]
```diff
@@ -166,28 +165,42 @@ public String getSQL(String cmd) {
...
+      if (code != 200) {
+        StringBuilder errorMessage = new StringBuilder("Failed : HTTP error code " + code);
+        logger.error("Failed to execute query: " + result);
+
+        KylinErrorResponse kylinErrorResponse = KylinErrorResponse.fromJson(result);
+        if (kylinErrorResponse == null) {
+          logger.error("Cannot get json from string: " + result);
+          // when code is 401, the response is html, not json
+          if (code == 401) {
+            errorMessage.append(". Error message: Unauthorized. This request requires "
```
Comment (yiming187) [inline kylin/src/main/java/org/apache/zeppelin/kylin/KylinInterpreter.java]: Could you move the period '.' into line 176. That will make the code easy to read. Same with line 188, and line 193.

**Comment by jinxliu** [comment]
@yiming187  I have made the changes and updated pull request. You can review now.
